src/Enrich.hs

```diff
@@ -83,18 +84,16 @@ mkEthEnvBounds vars = catMaybes $ mkBound <$> nub vars
       Nonce -> AbiUIntType 256
 
 -- | extracts bounds from the AbiTypes of Integer values in storage
-mkStorageBounds :: Store -> [Either StorageLocation StorageUpdate] -> [Exp Bool]
-mkStorageBounds store refs
-  = catMaybes $ mkBound <$> refs
+mkStorageBounds :: Store -> [Either StorageLocation StorageUpdate] -> [Exp Untimed Bool] -- is `Untimed` correct here?
```

Comment:
I may be wrong here, but I don't think so because preconditions are always untimed. It is only postconditions and returns which are timed. Remember that untimed doesn't mean "we don't care about the timing at all", it simply means "this expression makes no *syntactic* reference to the pre- and poststates". All expressions will always refer to some particular state semantically speaking, but imo that's the job of the backends to interpret. The role of the AST is to encode the syntax.

Reply:
hmmmm, I guess I kinda feel like preconditions have an implicit syntactic reference to the prestate? I think we could also probably reduce some duplication in the backends if we tag the expressions in the preconditinons as refering to the prestate?

Reply:
In that case no expressions should ever be untimed, and all of this timing stuff is completely unnecessary – we could just add a `When` to `TEntry`, which we forcibly insert in most cases, and be done with it. Which, yes, it does simplify things significantly.

I do see your point, and it's good, but let me offer a counterpoint: as this PR currently is, it helps us to guarantee, using Haskell's type system, that `pre` and `post` only ever occur where they're supposed to, because of the types of the fields in `Constructor` and `Behaviour`. If we said that `TEntry` always needs to contain a `When`, we wouldn't be certain that `preconditions :: [Exp Bool]` didn't contain any `post`s, and we couldn't use the target type `[Exp Untimed Bool]` to guide type checking.

That said, I do agree that backends shouldn't have to deal with figuring out which timing is correct for each particular untimed expression. So how about we add an additional refinement step, where we go e.g. `Behaviour -> RefinedBehaviour`, where all expressions in `RefinedBehaviour` are timed? Then we could forcibly insert `Pre` in all places which implicitly reference the prestate (and `Post` if there are any which reference the poststate, but I don't think there are). Then we would only ever expose `RefinedBehaviour` to the backends.

Reply:
True! Good call. So even if we do an extra refinement step we still need to expose both `Exp Timed` and `Exp Untimed` to the backends. But even then I think that the compilation process should at some point clearly distinguish between expressions which are implicitly timed and expressions which are explicitly timed. This allows us to more easily enforce the concrete syntax we have decided upon, and perhaps more importantly, makes it 100% clear that we actually are doing so.

Reply:
Or uh wait actually... couldn't each the extra refinement step also split each invariant into two timed expressions, one which points to the prestate and one which points to the poststate? This leaves less room for error in the backends because they can't do any faulty ``invariant `as` Pre`` or such, and again, we would only have to expose `Exp Timed` or some alias thereof, so backends would have less moving parts to care about.

Reply:
yeah I think that makes sense!

Reply:
Not sure it makes sense. Since at least for the SMT backend we use the expression only over the post state for the base case and then use it over the pre and post state for the inductive steps. 

Maybe we could have a `RefinedInvariant` type that exposes `assumption` (over Pre) and `postcondition` (over Post) fields or something?

Reply:
Hmm but that doesn't seem like it would pose a problem? Look at `mkInvariantQueries` in `SMT.hs`, specifically how we use the invariant expression [first in `mkInit`](https://github.com/ethereum/act/blob/21d4a60a8524c1ec84997547bbcd034da0326d35/src/SMT.hs#L247) and [then in `mkBehv`](https://github.com/ethereum/act/blob/21d4a60a8524c1ec84997547bbcd034da0326d35/src/SMT.hs#L272). We do indeed use the invariant expression as "two different" expressions: one which is `Post` (and also `Neg`, but that's SMT-specific I think?), and one which is `Pre` (but the latter isn't used in `mkInit`). What I'm suggesting is that we could do this duplication (or separation, depending on your viewpoint) before the code generation step, since this is literally what invariants *mean*. They are always two claims: one about the prestate, and then another about the poststate, and backends shouldn't be allowed to use them any other way.

Reply:
Yes!

Reply:
ok good, then in that case I agree 100% 😅